### PR TITLE
Adding support for ICON-EU in WPS

### PR DESCRIPTION
**Pull request by grafmi**
The ICON-EU grib2 data set is now able to be ingested by the ungrib program.
1. New Vtable:
   - A new Vtable file for ICON-EU (EU-Nest) has been created with the fields required by the
WRF model.
2. Information about the data:
   - The input data (grib2) of ICON-EU can be 
found here: https://opendata.dwd.de/weather/nwp/icon-eu/grib/ 
   - You will find the documentation about ICON and ICON-EU here: 
https://www.dwd.de/DWD/forschung/nwv/fepub/icon_database_main.pdf 
3. Modifications to the ungrib source code:
   - I added support for level code 150 in rd_grid2.F because model levels of ICON-EU are 
on generalized vertical height coordinates. The general vertical height coordinate is not 
specified by vertical coordinate parameters but by providing a 3D field, that specifies the 
geometric height of every model grid point in meters. For more details, see: 
http://www.cosmo-model.org/content/model/modules/coding/grib/gribVerticalCoordinates.pdf
4. Addition of new soil levels:
   - New soil levels from ICON-EU are added to METGRID.TBL.ARW for metgrid.exe. Details 
about the soil levels in ICON-EU can be found in the documentation (see above)
5. Conversion of soil moisture units to those expected by the WRF model:
   - The preprocessing in rrpr.F has to be modified to convert the units of column-integrated 
soil moisture provided by ICON-EU from kg m-2 to m3 m-2 for WRF input. 
## At the moment, I'm using a layer which exists only in ICON-EU as trigger to do the conversion. Maybe, there's a better solution. 

I have performed two model runs with exactly the same configuration and input from GFS 
and ICON-EU (init time 20200531_00 and lead time 72 h). The output of both runs are similar, 
but it can be recognized, e. g., that there is less shallow convection over Germany in the 
afternoon of the 31th May 2020. This was expected because GFS often shows too high 
low-level moisture values over Central Europe, especially in spring and summer.

GFS: https://www.meteoprime.ch/wrf_gfs/20200531_00/
ICON-EU: https://www.meteoprime.ch/wrf_icon/20200531_00/

### CODE REVIEW

File: ungrib/src/rd_grib2.F [hunk 1]
```diff
@@ -492,7 +492,7 @@ SUBROUTINE rd_grib2(junit, gribflnm, hdate,
               endif
 
               ! Scale lat/lon values to 0-180, default range is 1e6.
-              if (abs(map%lat1).ge.scale_factor) then
+              if (map%lat1.ge.scale_factor) then
```
Comment (davegill) [inline ungrib/src/rd_grib2.F:495]: @grafmi 
Michael,
Did we just change this and this is a revert?

Comment (mgduda) [reply ungrib/src/rd_grib2.F:495]: Yes, this seems to revert one of the changes made in PR #145 .

**Comment by davegill** [comment]
@grafmi 
Michael,
Would you provide a good commit message. For example, a few bullets for your modifications, and then a sentence or two for each bullet.

1. Since this is a pre-processor, it is difficult to gauge the correctness of the data you are using (assigning, etc). For your testing, do you have a sample **model** output with two different input data sources? Could you show those images (this input vs that input)? We need to know that the real.exe program is providing data correctly to the model with this new input. 
2. Also, if this is an available data set, in the commit message and in the Vtable, can you put the location of the online data source. An example for this type of information would be in the file WPS/ungrib/Variable_Tables/Vtable.GFS.

File: ungrib/src/rd_grib2.F (hunk 1 above)
Comment (grafmi) [reply ungrib/src/rd_grib2.F:495]: Dave, sorry, I thought that I'm working with the newest version, but it seems that it was version 4.1. This has to be reverted back to `if (abs(map%lat1).ge.scale_factor) then` Thanks!

Comment (grafmi) [reply ungrib/src/rd_grib2.F]: I reverted the accidental change back to the current version. So this issue should be fixed.

**Comment by grafmi** [comment]
@davegill 
Dave, I provide a better commit message. The first tests show that there was a problem with the soil moisture input (kg m-2), so I had to modify rrpr.F. Now, everything seems to work fine.

**Comment by davegill** [comment]
@grafmi
Michael,
I looked through the WRF-generated output fields. I could easily see the low-level moisture bias that you mentioned from the RH2m field. This looks pretty good to me.

Were any changes required in the WRF model to support this data? That would be a separate matter, but I am curious.

**Comment by davegill** [comment]
@mgduda @jimbresch 
Folks,
Would you review this PR. The WRF model output results seem to indicate that this was done correctly.

**Comment by davegill** [comment]
@mgduda @jimbresch 
I have highlighted an issue that the contributor posed about how to trigger the soil moisture conversion. Right now that conversion is based on the presence of a particular level.

**Comment by davegill** [comment]
@grafmi 
Michael,
1. Is the resolution of the data of the grib data about 6 km over Europe?
2. Is this data freely available in real-time?
3. Is access to the global 13-km data freely available also?
4. What files does a person download? There seem to be MANY files in the directory.

**Comment by smileMchen** [comment]
@grafmi 
I would like to run a test case using the ICON-EU data. Would you please specify what data I should download from the website you provided (https://opendata.dwd.de/weather/nwp/icon-eu/grib/)? I looked  at this website, there are many sub-directory and sub-subdirectory. Please clarify what data I should download from which sub-directory. Thanks.

**Comment by grafmi** [comment]
@davegill 
Dave, 
I didn't change anything at the WRF model at the moment. Currently, I don't understand completely how and where the interpolation of the soil moisture and temperature layers is done - I'm using the Noah Land Surface Model (option 2) with 4 layers. I will dive into the code in the next few days and check if there is room for further optimizations in the WRF model code. 

1. The resolution of the data is about 6.5 km on the native (triangular) grid or 0.06° on the regular lat-lon grid. 
2. Yes, the data are freely available in real-time. I use data on the regular lat-lon grid.
3. Yes, the global data are also freely available, but most necessary fields for WRF are on the native (triangular) grid. I don't know, whether WPS is able to handle that. 
4. @smileMchen I attached you a bash script that I'm using to download the necessary files. Hope that it helps. 
[download_icon.zip](https://github.com/wrf-model/WPS/files/4725528/download_icon.zip)

**Comment by smileMchen** [comment]
@grafmi 
Hi Michael, Thank you!

**Comment by sekluzia** [comment]
Hi, 

It would be great  also to include the global  ICON fields on the native (triangular) grid in the WPS.
 
Kind regards,
Artur

**Comment by mgduda** [comment]
The final result of the changes in the seven commits in this PR looks fine to me. Recognizing that this would create more work, and that we really should provide better documentation to help code contributors, I'd like to propose two changes:
1. In general, we'd like PRs to be created from a new branch, e.g., `ICON-support`, rather than from the `master` branch on a fork of the repository. We're moving toward the convention of including the name of the branch in the merge commit message, and it looks odd if the message says, e.g., that we're merging `master` onto `develop` for any reason except to include bug-fixes from a release branch. Committing changes to a new branch, and creating a PR from that, also helps contributors to keep the history of their `master` branches identical to the history of the main `master` branch; hard-resets can fix this, but having to hard-reset one's `master` branch suggests that an alternate development strategy might be better.
2. Unlike in the WRF repository, we don't "squash-and-merge" in the WPS repository, which means that the complete commit history is preserved in the repository. Consequently, it's nice to keep the commit history of branches as tidy as possible. For this PR, I think two commits should suffice: one to update ungrib, and a second to update the METGRID.TBL file.

**Comment by grafmi** [comment]
@mgduda As you proposed, I created another branch with the name ICON-support and reduced the number of commits from 7 to 4 for each modified file, which seems meaningful to me. If needed I can reduce the number of commits even further. Does I have to do another pull request to change the source branch from grafmi:master to grafmi:ICON-support?

**Comment by sfalmo** [comment]
Is this pull request still considered?

The possibility of ICON-EU based WRF runs for regions in Europe is certainly interesting. I think many users could benefit from this feature and would be glad to have it incorporated in a future release.

**Comment by weiwangncar** [comment]
@smileMchen Are you able to run a test with this dataset?

**Comment by smileMchen** [comment]
@weiwangncar 

I met some problems when ungribbing  the data.  The error message is like:

Subsoil level      0 in the GRIB2 file, was not found in the Vtable
Subsoil level    243 in the GRIB2 file, was not found in the Vtable
ERROR: Data not found: 2019-04-19_01:00:00.0000

However, the soil are included in the file downloaded from icon_eu. 
Vtable.ICON-EU also  looks fine. 

Ming

**Comment by weiwangncar** [comment]
@smileMchen Can you look into this a bit more, and work with @grafmi to resolve the problem? Thanks.

**Comment by smileMchen** [comment]
@weiwangncar 

I am using the master branch instead of the ICON-support branch, --- I believe the two branches are the same. But I will try the  ICON-support branch just in case.

**Comment by weiwangncar** [comment]
@smileMchen That would be good start. @grafmi did make changes to the code and Vtable.

**Comment by smileMchen** [comment]
@weiwangncar @grafmi
Wei,
ungrib still doesn't work, and I haven't figured out yet what is wrong. The datafile includes everything WPS requires. 
Michael,
I have made slight modification to the download script, basically to ensure one singe file only contains data for a single time. All others remain the same as in your original script. It is mysterious why ungrib cannot process the data I downloaded. Would you please send me one single-time ICON-EU datafile to take a look? Thanks.
Ming

**Comment by smileMchen** [comment]
@grafmi 
Michael,
Would you please send me one single  ICON-EU datafile to take a look? Or if you want to take a look at the  ICON-EU datafile I downloaded, please let me know. I can send you the file so that we can compare the difference. Thanks. 
Ming

**Comment by sfalmo** [comment]
I've had a look at this again and the following has all been done with the @grafmi:ICON-support branch.

I downloaded the ICON data with the script `download_icon.sh` from above (after creating and setting the directory for the data), ran `link_ungrib.csh` which produced a lot of `GRIBFILE.XXX` as expected and then started `ungrib.exe`. 

> ERROR: Data not found: 2019-04-19_01:00:00.0000

@smileMchen This error is most likely due to a wrong start or end date in `namelist.wps`, isn't it? At least I can avoid it that way, and get ungrib to finish successfully (but with warnings).

> Subsoil level 0 in the GRIB2 file, was not found in the Vtable
> Subsoil level 243 in the GRIB2 file, was not found in the Vtable

I am getting the same two warnings about superfluous soil levels as @smileMchen. But in my opinion, these can be ignored as they just point out that more soil levels are present in the grib files than ungrib is being told to process in the Vtable. Note that 6 levels for soil temperature and soil moisture are specified in [`Vtable.ICON-EU`](https://github.com/grafmi/WPS/blob/ICON-support/ungrib/Variable_Tables/Vtable.ICON-EU) but 7 are downloaded in `download_icon.sh` respectively.

One would either have to alter the Vtable to include the additional levels as well, or change the download script to not fetch them in the first place.
Btw: There are actually 8 levels for soil moisture and 9 levels for soil temperature as low as 20m on the OpenData server, see e.g. https://opendata.dwd.de/weather/nwp/icon-eu/grib/00/w_so/ and https://opendata.dwd.de/weather/nwp/icon-eu/grib/00/t_so/. @grafmi Why doesn't the Vtable and METGRID.TBL.ARW include all available levels?

Anyway, even with the currently used ones, I could continue with `metgrid.exe`, `real.exe` (after setting the values `num_metgrid_soil_levels = 6` and `num_metgrid_levels = 61` in `namelist.input`) and finally ran `wrf.exe` successfully.
The output parameters I looked at seemed alright, so I think with a bit of cleanup this feature could get merged very soon.

**Comment by dudhia** [comment]
The real.exe program can only handle specific soil levels from GFS, RUC and
ECMWF so far.

**Comment by weiwangncar** [comment]
@dudhia real should be able to handle any soil levels if they are set correctly in Vtable as well as METGRID.TBL. It looks like this PR has that.

**Comment by dudhia** [comment]
It has hardcoded names for ST00* etc., but maybe that has been changed.

**Comment by alexrsanchez** [comment]
Hi everyone,
I don't know if this is the right place to ask this question, but seemed logic to me as is about ICON-EU in WRF. Sorry in advance if it's not the proper place.
I am trying to run WRF using ICON-EU as input data instead of GFS. So far I've been able to edit the WPS properly in order to add the soil levels that ICON uses as described @grafmi in the first message of this conversation.

However, when running ./real.exe I got the following error:

> Using sfcprs to compute psfc
----- ERROR: The reference pressure is not monotonically decreasing
This tends to be caused by very high topography
(i,j) = 1 1 , topography = 0.0000000E+00 m
k = 1 , reference pressure = 100000.0 Pa
k = 2 , reference pressure = 100024.0 Pa
-------------- FATAL CALLED ---------------
FATAL CALLED FROM FILE: <stdin> LINE: 1236
In the dynamics namelist record, reduce etac from 9.9999998E-03
-------------------------------------------

As you know better than me, this is usually an error that raises with high topography, but as you can see, topography is 0m in my case. I am almost certain that the error raises because ICON vertical levels are reversed: as you can see in the attached image (WRF at the left part and ICON at the right), ICON assigns the first level (lev=1) to the top one and the last level (lev=61) to the surface layer, so pressure increases between levels 1 and 2 as the error points out, while GFS and WRF level assignment starts at the surface and ends at the highest desired level. How can I reverse the levels in the ICON grib2 files so I can remove this error?

WRF_vs_ICON_verticallevelspressure

**Comment by smileMchen** [comment]
@sfalmo @alexrsanchez 

I downloaded ICON data back in April using the script provided. I have the four datafile listed below:

icon_2021041900_000.grib2
icon_2021041900_001.grib2
icon_2021041900_002.grib2
icon_2021041900_003.grib2

All the above 4 files seem to have the same time, and ungrib is not able to find correct time contained in these files.
I can only tell their time based on the file name, i.e., 2021041900_000 (initialized time), 2021041900_001(1-hour FCTS), 2021041900_002(2-hour FCST), etc.

Would you please describe how you manage to make ungrib recognize the times in the field? 

I would appreciate if you can provide a sample download script, Vtable, and namelist.wps. I would like to at least ungrib these files first. 
 
Thanks in advance.

Ming

**Comment by sfalmo** [comment]
@smileMchen
Sure, I did the following:
1. Compile the ICON-support branch of @grafmi.
2. In the `WPS` directory: `mkdir -p data/icon data/tmp` and create the `download_icon.sh` script.
3. Run `./download_icon.sh <YYYYMMDD> <HH>`, so e.g. for today's 00Z data: `./download_icon.sh 20210630 00`. This takes some time, so you might want to run it in the background and observe the progress e.g. via `watch ls data/icon`. You should see this directory gradually populate with `icon_YYMMDDHH_XXX.grib2` files, where XXX is the forecast hour.
4. Run `./link_grib.csh data/icon`, which produces `GRIBFILE.AAA`, `GRIBFILE.AAB`, and so on.
5. Link the correct Vtable, i.e. `ln -s ungrib/Variable_Tables/Vtable.ICON-EU Vtable`. You see that I use the one provided by the ICON-support branch.
6. Make sure you have a valid `namelist.wps`, especially considering start_date and end_date. See below for the one I used.
7. Run `./ungrib.exe`, do not worry about warnings.
8. To proceed, you must have set up a domain with `geogrid.exe`. In my case, I just copied the `geo_em*` files from another setup (after I checked that the region is indeed covered by the ICON-EU model).
9. Link METGRID.TBL from the ICON-support branch, i.e. `ln -s metgrid/METGRID.TBL.ARW METGRID.TBL` and run `./metgrid.exe`.

And there you have it, a complete run of the WPS with ICON data.

Finally, here are the exact files I used:
[icon.zip](https://github.com/wrf-model/WPS/files/6743033/icon.zip)

**Comment by sfalmo** [comment]
@alexrsanchez 
And by the way, I do not get errors when running `real.exe` and `wrf.exe` after doing the WPS procedures from the previous comment.

Here's an example for the `namelist.input` I use:
[namelist.input.zip](https://github.com/wrf-model/WPS/files/6743450/namelist.input.zip)

**Comment by smileMchen** [comment]
@sfalmo 
Thank you very much for the kind information !
I am able to successfully run WRF using the ICON data as input.   Results are reasonable. 

@alexrsanchez
Would you please try WRFV4.3? I have attached a sample namelist.input just in case you want to take a look. 

Ming

[namelist.input.txt](https://github.com/wrf-model/WPS/files/6744524/namelist.input.txt)

**Comment by sekluzia** [comment]
@sfalmo

Will this work for the global ICON data (https://opendata.dwd.de/weather/nwp/icon/grib/)?

**Comment by sfalmo** [comment]
@smileMchen 
Good to know that you can reproduce a run with ICON-EU data now.

I had no success with the global ICON model due to a practical issue: I cannot find output on a regular lat/lon grid which is needed for ungrib. Only data on the native icosahedral grid seems to exist on the OpenData server (https://opendata.dwd.de/weather/nwp/icon/grib/).

This is strange because the official documents (https://www.dwd.de/DWD/forschung/nwv/fepub/icon_database_main.pdf, Chapters 9,10,11) state that all operational ICON models (global, EU and D2) should provide output on the native icosahedral as well as on a regular lat/lon grid (0.25° for global, 0.0625° for EU, 0.02° for D2). I only find the following on the OpenData server:
| | icosahedral | lat/lon |
| - | - | - |
| global ICON | X | |
| ICON-EU | | X |
| ICON-D2 | X | X |

Does anyone know what's up with the regular lat/lon gridded data of the global ICON model? If there is none, one must probably use some tool to regrid the icosahedral data themself, e.g. https://github.com/DeutscherWetterdienst/regrid.

**Comment by sekluzia** [comment]
> @smileMchen
> Good to know that you can reproduce a run with ICON-EU data now.
> 
> I had no success with the global ICON model due to a practical issue: I cannot find output on a regular lat/lon grid which is needed for ungrib. Only data on the native icosahedral grid seems to exist on the OpenData server (https://opendata.dwd.de/weather/nwp/icon/grib/).
> 
> This is strange because the official documents (https://www.dwd.de/DWD/forschung/nwv/fepub/icon_database_main.pdf, Chapters 9,10,11) state that all operational ICON models (global, EU and D2) should provide output on the native icosahedral as well as on a regular lat/lon grid (0.25° for global, 0.0625° for EU, 0.02° for D2). I only find the following on the OpenData server:
> 
> icosahedral	lat/lon
> global ICON	X	
> ICON-EU		X
> ICON-D2	X	X
> Does anyone know what's up with the regular lat/lon gridded data of the global ICON model? If there is none, one must probably use some tool to regrid the icosahedral data themself, e.g. https://github.com/DeutscherWetterdienst/regrid.

Thanks for your reply! It would be great to have an opportunity to run the WRF with the global ICON data so that the ICON data from the areas outside the European domain can be used in the WRF model as well.

**Comment by alexrsanchez** [comment]
Hi everyone, thanks a lot for your useful comments.
It turned out the problem to be the version of WRF I was using. I was having the errors with WPS4.1, but no errors when using WRFV4.3 as suggested by @smileMchen.

Regarding the lat-lon gridded data of global ICON model, it seems that there is limited data available at the DWD OpenData server, so maybe they run the model in that grid but that data is not open to the public...

**Comment by davegill** [comment]
@grafmi @alexrsanchez @sekluzia @smileMchen @sfalmo 
Developers and Contributors,
We are preparing for the next major WRF system release this spring, and we would like to include ICON-EU data support in WPS as part of the new capabilities.

I am trying to understand the current status. Here is what I _think_ the status is:

1. With global latitude/longitude data, the WPW v4.3 system (and later) is able to successfully process the data.
2. There does not seem to be a default location for the location of the global latitude/longitude data.
3. There is a tool to interpolate the icosahedral data to regular lat/lon.

**Comment by sfalmo** [comment]
Hello @davegill,

that's great news! It will certainly be appreciated, especially by all european WRF users.

The current status (at least from my view/experience) is the following:
- ICON-**EU** data (i.e. the european nest of ICON, not the global one) is provided on a regular lat/lon grid at a well-defined location on the OpenData server of the DWD.
- With the patches of this pull request by @grafmi, we can get WPS and WRF to run successfully with ICON-EU data. For downloading the latter, the above `download_icon.sh` script can be used. I could also find a similar script on the official website of the DWD (https://www.dwd.de/DE/leistungen/opendata/help/schluessel_datenformate/tools/fetchOpenData_sh.zip).

So I think we are good to go for the ICON-EU case, which this pull request is about. I assume that everything should also work for ICON-D2 data (an even higher-resolution nest for Germany), because you can find regular lat/lon gridded output on the OpenData server as well.

If you are thinking about **global** ICON data, there's still a technical issue. It is only available in its native icosahedral grid on the OpenData server - I don't know why, though. So you would have to use some tool to convert it to a regular lat/lon grid before giving it to WPS. There are some instructions (https://www.dwd.de/DE/leistungen/opendata/help/modelle/Opendata_cdo_EN.pdf) to do that. Also, the DWD provides a docker image (https://github.com/DeutscherWetterdienst/regrid) with which this transformation should be a one-liner. But this is all untested at the moment.

**Comment by meteolucania** [comment]
hi i'm trying to use wrf with icon data. I made the changes to the 4 files as indicated in the posts, used the icon vtable. everything crashesbal real.exe with the following error: https://pastebin.com/vNA5y8aR


nom I also understand why the metgrid level value is not present .... thanks a lot for the advice.

**Comment by WQadam** [comment]
Hi all,
Sorry this is abit late to the party as i have only just noticed this PR after forking my own which you can see at - https://github.com/WQadam/WPS This fork is still a work in progress in terms of model levels from the ICON model but i have this fork is up an running for icon-eu pressure levels and have tested the output. I can set up a seperate PR as i have in essence done the same changes but added an extra check to make WPS recognize DWD files which can then be set as an if statement when doing the soil moisture fix in ungrib.

Just wanted to try to get this thing moving again as i think it would be good to get DWD models supported by WRF.

A background to this is i initially started on this task in 2020 after a wrf forum post. https://forum.mmm.ucar.edu/threads/vtable-for-icon-eu.9391/ 
I then have only really got around to getting thing hard coded into WPS code so the user has to do minimal changes to the download of data and can just just the "raw" data from the server. 

Reading some of the comments i see stuff mentioned about soil levels with real.exe. Did anyone ever get to the bottom of real.exe needing to be changed? In my tests i use the noah lsm with 4 soil levels and seem to have reasonable soil moisture levels from what i can tell and the output of the model looks reasonable.

In terms of model levels - I have got this working using cdo to change the level type to hybrid and specifying the levels in the file from 1-n, again using a cdo command. My output seems reasonable as i did some tests vs the pressure level data. I can share a powerpoint i made back in 2020 to show the output if people would like to see it.

For the Global icon again i have got this working for a grid near Kenya. The user at the moment using the global icon needs to interpolate the downloaded files onto a latlon grid using cdo. Again, i can share instructions on this either in the Vtable or user guide or both.

Hope we can eventually get DWD models into WPS and wrf running from them soon

**Comment by weiwangncar** [comment]
@mgduda @smileMchen What do you think? Shall we ask WQadam to make a new PR for ICON-EU pressure level data? The advantage may be it will bring this PR up-to-date with newer WPS code. Also we seem to have lost contact with the original contributor somehow.

**Comment by grafmi** [comment]
> @mgduda @smileMchen What do you think? Shall we ask WQadam to make a new PR for ICON-EU pressure level data? The advantage may be it will bring this PR up-to-date with newer WPS code. Also we seem to have lost contact with the original contributor somehow.

Hello everybody, I'm still there, but unfortunately I had very few time for the contribution to this branch. I'm open to use the fork from @WQadam. Otherwise, I am happy to make minor modifications to this PR if the requirements are clearly defined. If I understood correctly, the following changes have to be done: 

- the branch has to updated so that it works with the latest WPS version v4.4 
- Vtable and METGRID.TBL.ARW should include all available soil levels (soil moisture and temperature)
- add an extra check to make WPS recognize DWD files as done by @WQadam

Is this correct or are there other necessary adjustments?

**Comment by weiwangncar** [comment]
@grafmi Thank you for the comments. It looks to me that those three items are needed to bring this PR up to date. Also the branch we are seeing is still the 'master'. Based on the comment by @mgduda, it should have a branch name that reflect the purpose of this PR.

**Comment by WQadam** [comment]
> @grafmi Thank you for the comments. It looks to me that those three items are needed to bring this PR up to date. Also the branch we are seeing is still the 'master'. Based on the comment by @mgduda, it should have a branch name that reflect the purpose of this PR.

@grafmi @weiwangncar 
It is them 3 items plus i have included a vtable for pressure levels aswell as model levels. The differing v tables is due to some variables not being included on pressure levels vs model levels so the pressure levels is done like the gfs variable but the model levels are done using the HRRR template of variables.
Note with the addition of WPS knowing DWD files then i use soil level 0 aswell as it being a dwd file as the indicator in met grid to convert SM correctly rather than just one level specific to the icon soil levels. I have added some explanation of soil levels from the DWD user guide in the readme file.

I have also added some readme files in my repository to hopefully give the user some explanation out of the box rather than having to search. At the moment i havent included any example download scripts but this could be included in the read me. 

One thing that my repository doesnt include is the change of unit in metgrid as the metgrid files produced still have the vtable units of kg/m2 rather than fraction in the metadata. Feel free to fix this if you think it is needed for clarity to the user.

Below i have done a quick comparison of the different icon outputs to check the model levels and pressure levels produce similar output data against the actual closest ascent that went up (assuming icon would feed this ascent at the data assimilation level).
icon comparison 121022

**Comment by sael9740** [comment]
Has there been any updates on this?

**Comment by WQadam** [comment]
My branched repository i believe is in a ready to go state (see my first comment on this thread) if everyone is alright ditching this current PR and me creating a separate PR to add the suite of icon models to WPS. I see madwrf are looking to add lines to the metgrid files in the next major release so not sure if they are getting the go ahead or we can jump in, as it might get abit messy trying to merge 2 pr's that are both changing the same file.

**Comment by grafmi** [comment]
Sorry, I forgot to mention that my changes for WPS 4.4 are also ready. The question is if I should do another PR or you would use the version of @WQadam. I made the following changes: 
- code is based upon the latest WPS version v4.4
- Vtable and METGRID.TBL.ARW include all available soil levels (soil moisture and temperature)
- extra check to make WPS recognize DWD files as done by @WQadam
- two Vtables for model and pressure levels

**Comment by weiwangncar** [comment]
@grafmi @WQadam Based on the review by @mgduda in June 2020, one has to create a new PR with more organized commits. So I hope one of you can do so quickly as we are planning on releasing 4.5 WPS between April 10-20. You can leave this PR here for the time being, and create a new one based on the latest WPS code. Thanks!

**Pull request closed without merging**